### PR TITLE
Change audit orchestration and implement `latest` strategy

### LINKED ISSUE #37: Change audit strategy
The current selection strategy for content auditing needs to be updated.  The priority for auditing should be as follows.

- Query `content_id` table and order by number of audits that have been performed (fewer first, more last)
- Secondary ordering by creation date `content_id.created_at` in descending order (newer first, older last)

This ensures that we focus on auditing the latest stuff first that has never been audited, and then once everything in the database has been audited, it starts focusing on re-auditing things with priority on the newest of those items.

### LINKED ISSUE #38: Add ability to run audits with different selection strategy
(low priority, only useful once the network is actually mostly working and most audits are successful)

depends on: https://github.com/pipermerriam/glados/issues/37

Lets call the strategy laid out in #37 the `latest` strategy, as it focuses on auditing the newest stuff.

We want two more strategies, and to modify `glados-audit` in any way necessary to allow us to run multiple audit processes concurrently (which might require database locking of some sort).

- A `random` strategy, that simply randomly audits things.
- A `missing` strategy that looks for failed audits and checks whether the data is still missing.

### PR DESCRIPTION
**Pull request by perama-v**
### Description

Updates the audit logic. This enables `glados-audit` to perform multiple audit strategies within the one binary. 
The PR includes the logic for the `latest` strategy and stubs for additional strategies.

Closes #37. 

### Dependencies

- Depends on #59.

### Overview

Note: Glados-monitor is unaffected. It continues to watch the chain tip of a non-portal node
and push content keys into the glados database. 

Multiple audit strategies are defined. Any (or all) may be invoked by spawning a new thread.
Each spawned thread sends audit tasks via a multi-producer single-consumer channel. Task-creating
threads do not alter the glados database.

An audit task is a content key from the the glados database that is expected to be in a portal node. Tasks are created according to different strategies:
- `latest` implemented
- `random` stub
- `failed` stub
- `<additional unimplemented strategies may be added later>`

A single thread watches the task channel and performs audits on anything received. Audit results
are stored in the glados database by this thread only, and no database locking is therefore required.

When `glados-audit`is started, all strategies are commenced. Any unimplemented strategy logs a warning.
A flag could later be added to chose a strategy/strategies. E.g., 
- No flag: Use all strategies
- Flag: Use comma-delimited value strategies only.

### Changes

- New `selection` module to handle the different strategies.
- New `SelectionStrategy` enum 
 
### Usage
```rs
let strategy = SelectionStrategy::Latest;
tokio::spawn(strategy.select_latest_content_for_audit(tx.clone(), conn.clone()));
```

### Testing
A test database generator is created that can be used for different strategies. See doctests for more. 

### Checklist

- [x] Add thread spawner for different strategies
- [x] Update latest with correct logic
- [x] Tidy commits
- [x] Add test for `latest` logic

### Not addressed
The following may be completed in a different PR(s).
- Add CLI
- Implement `failed` strategy
- Implement `random` strategy

### CODE REVIEW

File: glados-audit/src/generator.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+/// Finds and sends audit tasks for the 'latest' strategy.
+///
+/// Strategy achieved by:
+/// 1. Left joining contentkey table to the contentaudit table to find audits per key.
+/// 2. Filter for null audits (Exclude any item with an existing audit).
+/// 3. Sort ascending to have most recently added content keys first.
+async fn generate_latest(tx: mpsc::Sender<HistoryContentKey>, conn: DatabaseConnection) -> ! {
```
Comment (mrferris) [inline glados-audit/src/generator.rs:54]:
```suggestion
async fn select_latest_content_for_audit(tx: mpsc::Sender<HistoryContentKey>, conn: DatabaseConnection) -> ! {
```

File: glados-audit/src/generator.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        let body = HistoryContentKey::BlockBody(BlockBodyKey {
+            block_hash: hash.to_fixed_bytes(),
+        });
+
+        let receipts = HistoryContentKey::BlockReceipts(BlockReceiptsKey {
+            block_hash: hash.to_fixed_bytes(),
+        });
+
+        if let Err(e) = tx.send(header).await {
+            debug!(err=?e, "Could not send header for audit, channel might be closed.")
```
Comment (mrferris) [inline glados-audit/src/generator.rs:101]: The channel might also just be full, especially now that we're sending 30 queries per 120 seconds. 

This becomes a non-issue once we update Glados's `Client` library to run & wait on concurrent JSON RPC requests.

File: glados-audit/src/generator.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+/// A strategy is responsible for generating audit tasks.
+///
+/// An audit task is a content key from the the glados database that
+/// is expected to be in a portal node.
+pub enum Strategy {
+    /// Content that is:
+    /// 1. Not yet audited
+    /// 2. Sorted by date entered into glados database (newest first).
+    Latest,
+    /// Randomly generated content.
```
Comment (mrferris) [inline glados-audit/src/generator.rs:28]:
```suggestion
    /// Randomly selected content.
```

**Review by mrferris — CHANGES_REQUESTED** [review]
Looks good! 

Only thing I'd change are a few naming suggestions. The main one is that I think the name `selection.rs` describes what's being done here better than `generator.rs`. 

One suggestion for a future PR is that we include in the audit's DB record which selection strategy it was triggered by.

File: glados-audit/src/generator.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    Random,
+    /// Content that looks for failed audits and checks whether the data is still missing.
+    Missing,
```
[the remark this answers is not in the document]
Comment (perama-v) [reply glados-audit/src/generator.rs]: `Missing` is now `Failed`, and #62 tracks the new suggested strategies.

File: glados-audit/src/generator.rs (hunk 2 above)
Comment (perama-v) [reply glados-audit/src/generator.rs]: Updated the error message to be "full or closed" for now

**Comment by perama-v** [comment]
> One suggestion for a future PR is that we include within the audit's DB record which selection strategy it was triggered by.

See #63

**Comment by perama-v** [comment]
All suggested changes have been made

**Comment by pipermerriam** [comment]
can you rebase now that #59 is merged.

**Comment by perama-v** [comment]
Rebased on `master`

**Review by mrferris — APPROVED** [review]

**Comment by pipermerriam** [comment]
has conflicts now... probably from the other one I merged.

I'll get this merged and deployed tomorrow if the conflicts can get resolved by then.  Thanks again for all the great work.

**Comment by perama-v** [comment]
Rebase on `main` done 👍

**Comment by perama-v** [comment]
Testing of the `latest` strategy is now included.

**Pull request merged**